compose_reply: Include recipient link/mentions when forwarding messages.

PieterCK opened the pull request:
<!-- Describe your pull request here.-->
Changes (Updated on February 9, 2026):
- All "direct message" strings in message report template are now "message" only. As per discussed [here](https://chat.zulip.org/#narrow/channel/137-feedback/topic/Include.20Topic.20Name.20in.20quoted.20message.20.22said.22.20line/near/2328785).
- Forwarded channel messages now include a `with_url`  along with a topic link syntax to the original message.
- Forwarded direct messages now include the message's recipients.
- Reminders for channel message now include topic link syntax of the message, similar to forwarded channel message format.
- Reminders for direct message now include mentions of the message recipient, similar to forwarded direct message format.

The forwarded message's `with_url` and topic link syntax are not conditionally included only when the message is being sent to another topic -- like specified in #37103 --. This is because it seems like it's hard to implement, as discussed in [#frontend > Forward message template based on recipient #37145](https://chat.zulip.org/#narrow/channel/6-frontend/topic/Forward.20message.20template.20based.20on.20recipient.20.2337145/with/2361096).

Fixes part of #37103


**Screenshots and screen captures**:
(Re-tested on February 9, 2026) 
- Report direct message format:
  <img width="1085" height="341" alt="image" src="https://github.com/user-attachments/assets/acdf6b5d-f230-49c5-83b3-fd5a7dc286e1" />

- Quote message (Unchanged):
  <img width="1095" height="243" alt="image" src="https://github.com/user-attachments/assets/be83cdee-563a-4fd2-ae3a-8d40f576e084" />

- Forwarded public channel message:
  <img width="1085" height="210" alt="image" src="https://github.com/user-attachments/assets/c61f3076-fa87-43b4-af0b-3c763c786867" />

- Forwarded private channel message:
  <img width="1085" height="176" alt="image" src="https://github.com/user-attachments/assets/cf162242-97d1-4baf-a1bc-35997884f005" />

- Forwarded 1-on-1 direct message:
  <img width="1095" height="118" alt="image" src="https://github.com/user-attachments/assets/bc104367-e0b4-49a4-9d57-25865ff04ba0" />

- Forwarded one's own 1-on-1 direct message to others:
  <img width="1095" height="84" alt="image" src="https://github.com/user-attachments/assets/9e5fa8eb-42f0-445e-a02b-7dff532eeff5" />

- Forwarded direct message to self: 
  <img width="1085" height="146" alt="image" src="https://github.com/user-attachments/assets/d270df78-9a6b-424c-9205-ee3f20a892b0" />

- Forwarded direct message with 3 users:
  <img width="1085" height="121" alt="image" src="https://github.com/user-attachments/assets/27d4a4fd-feef-400c-a2d9-e0b16d86b07e" />

- Forwarded group direct message with 5 users:
  <img width="1085" height="210" alt="image" src="https://github.com/user-attachments/assets/4048105d-d98d-4d35-a1ac-2d88e0a36ee4" />

- Reminder for a direct message with 3 users:
  <img width="1048" height="202" alt="image" src="https://github.com/user-attachments/assets/3a36aa03-0f0f-4045-8601-e1fa320a1685" />

- Reminder for a direct message with 4 users:
  <img width="1048" height="264" alt="image" src="https://github.com/user-attachments/assets/f0478e9e-ad45-4058-8288-7f150e298257" />

- Reminder for a 1-on-1 direct message:
  <img width="1048" height="260" alt="image" src="https://github.com/user-attachments/assets/f318ef26-e5d0-45be-80be-abca605f2bd3" />

- Reminder for one's own direct message to someone:
  <img width="1048" height="264" alt="image" src="https://github.com/user-attachments/assets/f14e8cc8-94a3-428c-9554-1c8095fc431f" />

- Reminder for one's own direct message to oneself:
  <img width="1048" height="264" alt="image" src="https://github.com/user-attachments/assets/80d4253d-0dfa-425a-b797-c40d9f49e7c2" />
  (with note)
  <img width="1048" height="296" alt="image" src="https://github.com/user-attachments/assets/cc2518c5-47a2-4348-b227-332252ca8588" />
  (widgets)
  <img width="1113" height="398" alt="image" src="https://github.com/user-attachments/assets/fca4ef3a-0cbc-485f-8f7e-148bb76b4091" />

- Reminder for channel message:
  <img width="1048" height="293" alt="image" src="https://github.com/user-attachments/assets/5fbd430c-97d8-4c68-9559-be00226ab87a" />



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@karlstolley I'm putting this on your review queue for when it's been updated, but feel free to ask @amanagr to take over if that makes more sense.

PieterCK commented:
Updated `reminder.py` to keep its quoted direct message in synch with the new changes to forwarded direct message. Added a few more screenshorts in the description for this change.


This is ready for review

alya commented:
You can re-add the integration review label when ready.

alya commented:
> Do we want to also apply the same special format if these kinds of messages are sent to one's self? e.g. You sent yourself a poll.

Yeah, I think that will be nice for todo lists (probably not uncommon) and polls (presumably rare, but still reasonable for completeness?).

PieterCK commented:
> Yeah, I think that will be nice for todo lists (probably not uncommon) and polls (presumably rare, but still reasonable for completeness?).

Ok, updated the format for these types of reminders too.

Implementation-wise, I re-wrote basically all the new changes in `reminder.py`. The previous approach in this PR would produce untranslatable strings (or hard to understand at least), since I was appending strings marked for translation in a hacky way. Would it make sense to mark this as ready for integration review, or should this go through at least another round of maintainer review again?

alya commented:
I guess let's have @amanagr take one more look.

amanagr commented on zerver/lib/reminders.py:
```diff
@@ -82,9 +84,18 @@ def get_reminder_formatted_content(
+            if user["id"] is not message.sender.id
+        ]
+        if not recipients:
+            recipients = [message.sender]
```
This can be extracted into a function to simplify the code reading experience and possibly be used in other places.

amanagr replied:
I see that your next commit makes good use of this, so we can skip that here.

amanagr commented:
Left a couple minor comments. Changes look good to me otherwise. Tested manually as well.

timabbott commented on web/tests/compose_actions.test.cjs:
```diff
@@ -570,9 +591,28 @@ test("quote_message", ({disallow, override, override_rewire}) => {
+        ),
+        near_url: hash_util.by_conversation_and_time_url(selected_message),
+        fence: fenced_code.get_unused_fence("Testing."),
+        message: "Testing.",
```
This isn't a good testing habit. The `quote_message_template` function is fine, but I think you want to put the actual concrete values here in these parameters, rather than calling functions to compute them, since there's a good chance a bug that impacts this output would be in one of the functions you're calling here.

timabbott replied:
This is certainly a ways better.

timabbott commented on zerver/lib/message.py:
```diff
@@ -1810,3 +1810,16 @@ def is_message_to_self(message: Message) -> bool:
+    other_users: str = ", ".join(recipient_list)
+    if len(recipient_list) > 1:
+        other_users += ","
+    return f"{other_users} and {last_user}"
```
This is invalid internationalization -- the "and" here isn't tagged for translation. 

That's why we had these strings done carefully to include the "and" part in what is tagged for translation.

timabbott commented:
I made the tweaks noted above; I think this is a fine checkpoint that basically completes the issue. 

Can you bump the chat.zulip.org conversations regarding what you believe remains for #37103? I think it's possible we want to just close it and maybe file some follow-ups. The only remaining issues I'm aware of are:
- The reminder syntax for the original topic doesn't have way to the specific message ID in question when linking to the topic, and thus maybe not producing the optimal /with/ link when describing the channel/topic (I think it might make one, but with a random other message ID in that topic in the /with/ link?). This probably deserves its own chat.zulip.org thread.
- We always include the destination if you clicked "forward message", even if you end up selecting as the compose recipient for forwarding the original conversation you're replying to, which seems basically Fine and consistent with the original issue write-up. But maybe I'm missing some detail from the previous discussions? Best to engage in CZO for better discussion and we can close the issue as completed if we conclude it's warranted.

PieterCK commented:
> The reminder syntax for the original topic doesn't have way to the specific message ID in question when linking to the topic, and thus maybe not producing the optimal /with/ link when describing the channel/topic (I think it might make one, but with a random other message ID in that topic in the /with/ link?). This probably deserves its own chat.zulip.org thread.

Filed #38089 to track this, it looks like the /with/ link will point to the latest message in that topic at the time the reminder is set and not random.
